les/costtracker.go

```diff
@@ -193,26 +232,47 @@ func (ct *costTracker) gfLoop() {
 	}
 	gf := gfSum / gfWeight
 	ct.gf = gf
+	totalRecharge := ct.utilTarget * gf
 	ct.gfUpdateCh = make(chan gfUpdate, 100)
 
 	go func() {
+		saveCostFactor := func() {
+			var data [16]byte
+			binary.BigEndian.PutUint64(data[0:8], math.Float64bits(gfSum))
+			binary.BigEndian.PutUint64(data[8:16], math.Float64bits(gfWeight))
+			ct.db.Put([]byte(gfDbKey), data[:])
+			log.Debug("global cost factor saved", "sum", time.Duration(gfSum), "weight", time.Duration(gfWeight))
+		}
+		saveTicker := time.NewTicker(time.Minute * 10)
+
 		for {
 			select {
 			case r := <-ct.gfUpdateCh:
 				now := mclock.Now()
 				max := r.servingTime * gf
+				if ct.logRelCost != nil && r.avgTime > 1e-20 {
+					ct.logRelCost.Update(max / r.avgTime)
+				}
+				if r.servingTime > 1000000000 {
+					ct.logger.Event(fmt.Sprintf("Very long servingTime = %f  avgTime = %f  costFactor = %f", r.servingTime, r.avgTime, gf))
+				}
+				if max > r.avgTime*maxCostFactor {
+					max = r.avgTime * maxCostFactor
+					r.servingTime = max / gf
+				}
 				if r.avgTime > max {
 					max = r.avgTime
 				}
 				dt := float64(now - expUpdate)
 				expUpdate = now
 				gfUsage = gfUsage*math.Exp(-dt/float64(gfUsageTC)) + max*1000000/float64(gfUsageTC)
 
-				if gfUsage >= gfUsageThreshold*ct.utilTarget*gf {
+				if gfUsage >= gfUsageThreshold*totalRecharge {
 					gfSum += r.avgTime
 					gfWeight += r.servingTime
 					if time.Duration(now-lastUpdate) > time.Second {
 						gf = gfSum / gfWeight
```

Comment:
Let's choose a better way in the future.
In this formula, we will calculate the `factor` based on the accumulated `gfWeight`. So it's hard to decrease `factor` from high value.